internal/pkg/cli/local_run.go

```diff
@@ -0,0 +1,336 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/ssm"
```

Comment:
Can we keep the dependency on `aws-sdk-go/service` within `internal/aws`, so that `internal/cli` depends on `internal/aws`, and `internal/aws` depends on `aws-sdk-go/service`?


```go
// In internal/aws/ssm/ssm.go
func (s *SSM) GetSecretValue(name string) (string, error)

// In internal/aws/secretsmanager/secretsmanager.go
func (s *SecretsManager) GetSecretValue(name string) (string, error)

// Here in `getDecrpyted...`
value, err := o.secretsmanager.GetSecretValue(name)
```

A few benefits that I can think of:
1. This keeps most, if not all, of external dependencies within fewer packages. Consider if `aws-sdk-go/service` changes their API, we'd only need to touch our codes in `internal/aws`, instead of all over the place in the whole codebase.
2. Keeps logic that's specific to interacting with `aws-sdk-go/service` within `internal/aws`. For example, pagination is a common thing in HTTP requests. Consider this - if each of our packages directly interacts with `aws-sdk-go` instead of through `internal/aws`, we'd have paginations for-loops all over the place, distracting us from the main logic, making the code harder to read. 
3. Reuse the code. Still the pagination example - if we had handled the pagination in a `internal/aws` package, we wouldn't have to have the same pagination for-loop every where. 
4. Our internal packages can benefit from our own abstraction.  Consider this example - suppose we have
```go
// In internal/aws/ecs/ecs.go

import "github.com/aws/aws-sdk-go/service/ecs"

type ecsClient interface {
   DeployService(*ecs.DeployServiceInput) error
}

type ECS struct {
   client ecsClient
}

func New(sess session) *ECS { 
    return &ECS{ 
        client: ecs.New(sess) // initialized with a client from aws-sdk-go/ecs
    }
}

func (ecs *ECS) UpdateService() error {
     if err := ecs.client.UpdateService(&ecs.DeployServiceInput{}); err != nil {...}
}
```
In a third package (e.g. `cli`), we can have ⬇️ :
```go
import "github.com/aws/copilot-cli/internal/pkg/aws/ecs"

type  svcUpdator interface {
   UpdateService() error
}

type lbws struct {
   updator svcUpdator
}

func NewLBWS() *lbws {
  return &lbws{
     deployer: ecs.New(sess) // This is a client from internal ecs package, not aws-sdk-go.
  }
}

func (lbws *lbws) Execute() error {
   if err := lbws.deployer.UpdateService(); err != nil { ... }
}
```

If some day, we decide to switch from ECS to AppRunner as the service provider, we can just add a new package:
```go
// In internal/aws/apprunner/apprunner.go
import "github.com/aws/aws-sdk-go/service/apprunner"

type apprunnerClient interface {
   DeployService(*apprunner.DeployServiceInput) error // Unlike aws-sdk-go/service/ecs, the apprunner API is `DeployService`, not `UpdateService`.
}

type AppRunner struct {
   client apprunnerClient // This will be initialized with apprunner.New(sess), that is, this is a client from aws-sdk-go
}

func (ar *AppRunner) UpdateService() error { // We abstract away `DeployService`, and keep the same `UpdateService` interface as our `internal/aws/ecs` package to minimize the amount of code changes
     if err := ar.client.DeployService(&apprunner.DeployServiceInput{}); err != nil {...}
}
```
and then just change one line of code in `cli` 
```go
func NewLBWS() *lbws {
  return &lbws{
     deployer: apprunner.New(sess) // Previously ecs.New(sess). 
  }
}
```


This is easy and clean. However, consider if we didn't have `internal/aws` as a layer between `ask-sdk-go` and `cli`.......  we'd have a lot of code changes all over the place:
```go
// In cli package
import "github.com/aws/aws-sdk-go/service/apprunner" // changed from import "github.com/aws/aws-sdk-go/service/ecs" 

type  svcUpdator interface {
   DeployService(*apprunner.DeployServiceInput) error // changed from UpdateService(*ecs.UpdateServiceInput)
}

type lbws struct {
   updator svcUpdator
}

func NewLBWS() *lbws {
  return &lbws{
     deployer: apprunner.New(sess) // changed from ecs.New(sess)
  }
}

func (lbws *lbws) Execute() error {
   if err := lbws.deployer.DeployService(&apprunner.DeployServiceInput); err != nil {  // changed from  lbws.deployer.UpdateService(&ecs.UpdateServiceInput)
     // A bunch of error handling specific to apprunner.DeployService // changed from "a bunch of error handling specific to ecs.UpdateService
   }
}
```


Hope this helps!

Reply:
Makes sense. This is some explanation👀.